### ISSUE TITLE
Inconsistent handling of warnings with log-level and stderr

### ISSUE THREAD

**Issue by phiggins** [inspec/inspec#3648]
### 🎛 Description

This is a bug report.

Occasionally I find profiles that make inspec generate warnings. This causes problems when we're expecting inspec to generate parseable json, for example. The bug I'm experiencing is that the various inspec commands seem to inconsistently honor the `--log-level` flag and printing warnings to stderr, both of which I could use to workaround the issue without changes to inspec.

### 🌍 InSpec and Platform Version

```
vagrant@ubuntu-xenial:~$ inspec --version
3.0.61
vagrant@ubuntu-xenial:~$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 16.04.5 LTS
Release:        16.04
Codename:       xenial
```

### 🤔 Replication Case

* archive does not honor log-level or output warnings to stderr:
    ```
    vagrant@ubuntu-xenial:~$ inspec archive warning-profile
    Dependencies for profile warning-profile successfully vendored to /home/vagrant/warning-profile/vendor
    I, [2018-12-05T01:09:16.603213 #9955]  INFO -- : Checking profile in warning-profile
    I, [2018-12-05T01:09:16.603701 #9955]  INFO -- : Metadata OK.
    [2018-12-05T01:09:16+00:00] WARN: Overwriting resource shadow. To reference a specific version of shadow use the resource() method
    I, [2018-12-05T01:09:16.606544 #9955]  INFO -- : Found 1 controls.
    I, [2018-12-05T01:09:16.606888 #9955]  INFO -- : Control definitions OK.
    I, [2018-12-05T01:09:16.607359 #9955]  INFO -- : Archive /home/vagrant/warning-profile-0.1.0.tar.gz exists already. Use --overwrite.
    vagrant@ubuntu-xenial:~$ inspec archive warning-profile --log-level=fatal
    Dependencies for profile warning-profile successfully vendored to /home/vagrant/warning-profile/vendor
    [2018-12-05T01:08:53+00:00] WARN: Overwriting resource shadow. To reference a specific version of shadow use the resource() method
    vagrant@ubuntu-xenial:~$ inspec archive warning-profile 2>/dev/null
    Dependencies for profile warning-profile successfully vendored to /home/vagrant/warning-profile/vendor
    I, [2018-12-05T01:09:08.432305 #9953]  INFO -- : Checking profile in warning-profile
    I, [2018-12-05T01:09:08.432959 #9953]  INFO -- : Metadata OK.
    [2018-12-05T01:09:08+00:00] WARN: Overwriting resource shadow. To reference a specific version of shadow use the resource() method
    I, [2018-12-05T01:09:08.436225 #9953]  INFO -- : Found 1 controls.
    I, [2018-12-05T01:09:08.436551 #9953]  INFO -- : Control definitions OK.
    I, [2018-12-05T01:09:08.436948 #9953]  INFO -- : Archive /home/vagrant/warning-profile-0.1.0.tar.gz exists already. Use --overwrite.
    ```

* check does not honor log-level or output warnings to stderr:
    ```
    vagrant@ubuntu-xenial:~$ inspec check warning-profile --format=json 2>/dev/null
    [2018-12-05T01:04:19+00:00] WARN: Overwriting resource shadow. To reference a specific version of shadow use the resource() method
    {"summary":{"valid":true,"timestamp":"2018-12-05T01:04:19+00:00","location":"warning-profile","profile":"warning-profile","controls":1},"errors":[],"warnings":[]}
    vagrant@ubuntu-xenial:~$ inspec check warning-profile --format=json --log-level=fatal
    [2018-12-05T01:04:30+00:00] WARN: Overwriting resource shadow. To reference a specific version of shadow use the resource() method
    {"summary":{"valid":true,"timestamp":"2018-12-05T01:04:30+00:00","location":"warning-profile","profile":"warning-profile","controls":1},"errors":[],"warnings":[]}
    ```

* exec honors log-level but does not output warnings to stderr:
    ```
    vagrant@ubuntu-xenial:~$ inspec exec warning-profile
    [2018-12-05T01:12:24+00:00] WARN: Overwriting resource shadow. To reference a specific version of shadow use the resource() method

    Profile: A profile that generates a warning (warning-profile)
    Version: 0.1.0
    Target:  local://

      ↺  this generates a deprecation warning: Using shadow generates a warning
         ↺  Can't read file: /etc/shadow


    Profile Summary: 0 successful controls, 0 control failures, 1 control skipped
    Test Summary: 0 successful, 0 failures, 1 skipped
    vagrant@ubuntu-xenial:~$ inspec exec warning-profile --log-level=fatal

    Profile: A profile that generates a warning (warning-profile)
    Version: 0.1.0
    Target:  local://

      ↺  this generates a deprecation warning: Using shadow generates a warning
         ↺  Can't read file: /etc/shadow


    Profile Summary: 0 successful controls, 0 control failures, 1 control skipped
    Test Summary: 0 successful, 0 failures, 1 skipped
    vagrant@ubuntu-xenial:~$ inspec exec warning-profile 2>/dev/null
    [2018-12-05T01:12:47+00:00] WARN: Overwriting resource shadow. To reference a specific version of shadow use the resource() method

    Profile: A profile that generates a warning (warning-profile)
    Version: 0.1.0
    Target:  local://

      ↺  this generates a deprecation warning: Using shadow generates a warning
         ↺  Can't read file: /etc/shadow


    Profile Summary: 0 successful controls, 0 control failures, 1 control skipped
    Test Summary: 0 successful, 0 failures, 1 skipped
    ```

* json honors log-level and outputs warnings to stderr:
    ```
    vagrant@ubuntu-xenial:~$ inspec json warning-profile
    [2018-12-05T01:06:32+00:00] WARN: Overwriting resource shadow. To reference a specific version of shadow use the resource() method
    {"name":"warning-profile","title":"A profile that generates a warning","maintainer":"The Authors","copyright":"The Authors","copyright_email":"you@example.com","license":"Apache-2.0","summary":"An InSpec Compliance Profile","version":"0.1.0","supports":[],"controls":[{"title":"Using shadow generates a warning","desc":"Using shadow generates a warning","descriptions":{"default":"Using shadow generates a warning"},"impact":0.5,"refs":[],"tags":{},"code":"control \"this generates a deprecation warning\" do\n  title \"Using shadow generates a warning\"\n  desc \"Using shadow generates a warning\"\n  describe shadow.where { user  !~ /.+/ } do\n    its(\"raw_data\") { should be_empty }\n  end\nend\n","source_location":{"ref":"warning-profile/controls/controls.rb","line":16},"id":"this generates a deprecation warning"}],"groups":[{"title":null,"controls":["this generates a deprecation warning"],"id":"controls/controls.rb"}],"attributes":[],"sha256":"8a1b089a41ec1365de4d3d09fc0c8b048785eaa46791ab7806e4b4cf26fe2a81","status":"loaded","generator":{"name":"inspec","version":"3.0.61"}}
    vagrant@ubuntu-xenial:~$ inspec json warning-profile --log-level=fatal
    {"name":"warning-profile","title":"A profile that generates a warning","maintainer":"The Authors","copyright":"The Authors","copyright_email":"you@example.com","license":"Apache-2.0","summary":"An InSpec Compliance Profile","version":"0.1.0","supports":[],"controls":[{"title":"Using shadow generates a warning","desc":"Using shadow generates a warning","descriptions":{"default":"Using shadow generates a warning"},"impact":0.5,"refs":[],"tags":{},"code":"control \"this generates a deprecation warning\" do\n  title \"Using shadow generates a warning\"\n  desc \"Using shadow generates a warning\"\n  describe shadow.where { user  !~ /.+/ } do\n    its(\"raw_data\") { should be_empty }\n  end\nend\n","source_location":{"ref":"warning-profile/controls/controls.rb","line":16},"id":"this generates a deprecation warning"}],"groups":[{"title":null,"controls":["this generates a deprecation warning"],"id":"controls/controls.rb"}],"attributes":[],"sha256":"8a1b089a41ec1365de4d3d09fc0c8b048785eaa46791ab7806e4b4cf26fe2a81","status":"loaded","generator":{"name":"inspec","version":"3.0.61"}}
    vagrant@ubuntu-xenial:~$ inspec json warning-profile 2>/dev/null
    {"name":"warning-profile","title":"A profile that generates a warning","maintainer":"The Authors","copyright":"The Authors","copyright_email":"you@example.com","license":"Apache-2.0","summary":"An InSpec Compliance Profile","version":"0.1.0","supports":[],"controls":[{"title":"Using shadow generates a warning","desc":"Using shadow generates a warning","descriptions":{"default":"Using shadow generates a warning"},"impact":0.5,"refs":[],"tags":{},"code":"control \"this generates a deprecation warning\" do\n  title \"Using shadow generates a warning\"\n  desc \"Using shadow generates a warning\"\n  describe shadow.where { user  !~ /.+/ } do\n    its(\"raw_data\") { should be_empty }\n  end\nend\n","source_location":{"ref":"warning-profile/controls/controls.rb","line":16},"id":"this generates a deprecation warning"}],"groups":[{"title":null,"controls":["this generates a deprecation warning"],"id":"controls/controls.rb"}],"attributes":[],"sha256":"8a1b089a41ec1365de4d3d09fc0c8b048785eaa46791ab7806e4b4cf26fe2a81","status":"loaded","generator":{"name":"inspec","version":"3.0.61"}}
    ```

**Comment by jerryaldrichiii** [comment]
I believe the issue is `warn` vs `Inspec::Log.warn`. The former will not be affected by `--log-level`. We should do a search and replace throughout the code base in my opinion.

**Comment by clintoncwolfe** [comment]
InSpec code is all over the place on this, calling bare `puts` in places, using Inspec::Log in others, and `warn` in other places.

I am very much in favor of fixing this, but note this will also involve a lot of corrections in the tests, since they current looks for errors in the "wrong" locations.

**Comment by clintoncwolfe** [comment]
Refs #874

**Comment by phiggins** [comment]
#3972 #4001

**Comment by zenspider** [comment]
Fixed by #4011.
